types.go

```diff
@@ -242,20 +242,176 @@ type RaftChanges struct {
+		if v.IsCurrentVoter() == isCurrentVoter {
+			servers[id] = v
+		}
+	}
+	return servers
+}
+
+type CategorizedServers struct {
+	// StaleNonVoters are the IDs of non-voting server nodes in the raft configuration
+	// that are not present in the delegates view of the server nodes should be available
+	StaleNonVoters RaftServerEligibility
+	// StaleVoters are the IDs of voting server nodes in the raft configuration
+	//that are not present in the delegates view of the servers node should be available
+	StaleVoters RaftServerEligibility
+	// FailedNonVoters are the IDs of non-voting server nodes in the raft cluster
+	// that the delegate has indicated are in a failed state
+	FailedNonVoters RaftServerEligibility
+	// FailedVoters are the IDs of voting server nodes in the raft cluster
+	// that the delegate has indicated are in a failed state
+	FailedVoters RaftServerEligibility
+	// HealthyNonVoters are the IDs of non-voting server nodes
+	// that the delegate has indicated are operating as intended
+	HealthyNonVoters RaftServerEligibility
+	// HealthyVoters are the IDs of voting server nodes
+	// that the delegate has indicated are operating as intended
+	HealthyVoters RaftServerEligibility
+}
+
+// PotentialVoters sums the number of servers that have the potential to become voters.
+func (s *CategorizedServers) PotentialVoters() int {
+	potentialVoters := 0
+
+	for _, v := range s.FailedNonVoters {
+		if v.IsPotentialVoter() {
+			potentialVoters++
+		}
+	}
+
+	for _, v := range s.FailedVoters {
+		if v.IsPotentialVoter() {
+			potentialVoters++
+		}
+	}
+
+	for _, v := range s.HealthyNonVoters {
+		if v.IsPotentialVoter() {
+			potentialVoters++
+		}
+	}
+
+	for _, v := range s.HealthyVoters {
+		if v.IsPotentialVoter() {
+			potentialVoters++
+		}
+	}
+
+	return potentialVoters
+}
+
+// FailedServers is essentially a DTO to support the promoter interface
 type FailedServers struct {
-	// StaleNonVoters are the ids of those server in the raft configuration as non-voters
-	// that are not present in the delegates view of what servers should be available
-	StaleNonVoters []raft.ServerID
-	// StaleVoters are the ids of those servers in the raft configuration as voters that
-	// are not present in the delegates view of what servers should be available
-	StaleVoters []raft.ServerID
-
-	// FailedNonVoters are the servers without voting rights in the cluster that the
-	// delegate has indicated are in a failed state
+	StaleNonVoters  []raft.ServerID
+	StaleVoters     []raft.ServerID
 	FailedNonVoters []*Server
-	// FailedVoters are the servers without voting rights in the cluster that the
-	// delegate has indicated are in a failed state
-	FailedVoters []*Server
+	FailedVoters    []*Server
+}
+
+// convertToFailedServers uses CategorizedServers to create the FailedServers
+// struct which can be used to maintain compatibility with the promoter interface
+func (s *CategorizedServers) convertToFailedServers(state *State) *FailedServers {
```

Comment:
Can we avoid the need for conversion, and instead of introducing CategorizedServers just augment FailedServers?  After all, the two structs are very similar.  And it seems odd to make every field (stale nonvoters, stale voters, etc) contain the redundant information about eligibility.  Couldn't you instead just have a single map from serverid->VoterEligibility that contains all nodes, instead of making each node list be a map?